parallel.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2023 Stock Parfait
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"sync"
+)
+
+type contextKey int
+
+const (
+	parallelKey contextKey = iota
+)
+
+// TestSerialize forces the number of workers in ParallelMap to be 1, thereby
+// running jobs serially and in the strict FIFO order. This helps make tests
+// deterministic.
+func TestSerialize(ctx context.Context) context.Context {
+	return context.WithValue(ctx, parallelKey, true)
+}
+
+func isSerialized(ctx context.Context) bool {
+	v, ok := ctx.Value(parallelKey).(bool)
+	return ok && v
+}
+
+// Job is a unit of work to be done, which returns a result.
+type Job[T any] func() T
+
+type parallelMapIter[T any] struct {
+	ctx     context.Context  // potentially cancelable context
+	workers int              // maximum number of parallel jobs allowed
+	jobs    int              // number of jobs currently running
+	it      Iterator[Job[T]] // job iterator
+	resCh   chan T           // workers send their results to this channel
+	done    bool
+	mux     sync.Mutex // to make Next() go routine safe
+}
+
+var _ Iterator[int] = &parallelMapIter[int]{}
+
+// ParallelMap runs multiple jobs in parallel on a given number of workers,
+// 0=unlimited, collects their results and returns as an iterator. The order of
+// the results is not guaranteed, unless the number of workers is 1.
+//
+// Canceling the supplied context immediately stops queuing new jobs, but the
+// jobs that already started will finish and their results will be returned.
+// Therefore, it is important to flush the iterator after canceling the context
+// to release all the resources.
+//
+// No job is started by this method itself. Jobs begin to run on the first
+// Next() call on the result iterator, which is go routine safe.
+//
+// Example usage:
+//
+//	m := ParallelMap(context.Background(), 2, jobsIter)
+//	for v, ok := m.Next(); ok; v, ok = m.Next() {
+//	  // Process v
+//	}
+func ParallelMap[T any](ctx context.Context, workers int, jobs Iterator[Job[T]]) Iterator[T] {
```

Comment:
Change signature to match `Map`. The current signature can be modeled as an instance of `ParallelMap(ctx, workers, iter, f)` by making `f(job)` call `job` and return its result.